dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java

```diff
@@ -62,18 +63,59 @@
     public static final String ERROR_PARSING_WORKFLOW_YOU_MAY_HAVE_A_RECURSIVE_IMPORT = "Error parsing workflow. You may have a recursive import.";
     private static final Pattern IMPORT_PATTERN = Pattern.compile("^import\\s+\"(\\S+)\"");
 
+    public static void checkForRecursiveLocalImports(String content, Set<SourceFile> sourceFiles, Set<String> absolutePaths, String parent)
+            throws Exception {
+        // Use matcher to get imports
+        String[] lines = StringUtils.split(content, '\n');
+        for (String line : lines) {
+            Matcher m = IMPORT_PATTERN.matcher(line);
+
+            while (m.find()) {
+                String match = m.group(1);
+                if (!match.startsWith("http://") && !match.startsWith("https://")) { // Don't resolve URLs
+                    String localRelativePath = match.replaceFirst("file://", "");
+                    String absolutePath = LanguageHandlerHelper.convertRelativePathToAbsolutePath(parent, localRelativePath);
+                    if (absolutePaths.contains(absolutePath)) {
+                        throw new Exception("Recursive local import detected: " + absolutePath);
+                    }
+                    // Creating a new set to avoid false positive caused by multiple "branches" that have the same import
+                    Set<String> newAbsolutePaths = new HashSet<>();
+                    newAbsolutePaths.addAll(absolutePaths);
+                    newAbsolutePaths.add(absolutePath); // remove file:// from path
+                    Optional<SourceFile> sourcefile = sourceFiles.stream()
+                            .filter(sourceFile -> sourceFile.getAbsolutePath().equals(absolutePath)).findFirst();
+                    if (sourcefile.isPresent()) {
+                        File file = new File(absolutePath);
+                        String newParent = file.getParent();
+                        checkForRecursiveLocalImports(sourcefile.get().getContent(), sourceFiles, newAbsolutePaths, newParent);
+                    }
+                }
+            }
+        }
+
+    }
+
     @Override
     public Version parseWorkflowContent(String filepath, String content, Set<SourceFile> sourceFiles, Version version) {
+        try {
+            File file = new File(filepath);
+            String parent = file.getParent();
+            checkForRecursiveLocalImports(content, sourceFiles, new HashSet<>(), parent);
+        } catch (Exception e) {
+            LOG.error("Recursive local imports found: " + version.getName(), e);
```

Comment:
If I understand correctly, any exception thrown from checkForRecursiveLocalImports will result in a log message saying that recursive imports were found. This is misleading if, for example, there is a null pointer exception or a IllegalArgumentException this will still catch it and give an odd validation message too. 

Make a custom exception https://www.baeldung.com/java-new-custom-exception

Reply:
See initial PR comment, still trying to think of a non-raw exception to use before resorting to custom exception